src/crypto/tls/metrics_test.go

```diff
@@ -0,0 +1,139 @@
+// Copyright 2020 Cloudflare, Inc. All rights reserved. Use of this source code
+// is governed by a BSD-style license that can be found in the LICENSE file.
+
+package tls
+
+import (
+	"crypto/x509"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"testing"
+	"time"
+)
+
+func (e EXP_EventTLS13ServerHandshakeTimingInfo) isMonotonic() bool {
+	return e.ReadClientHello <= e.WriteServerHello &&
+		e.WriteServerHello <= e.WriteEncryptedExtensions &&
+		e.WriteEncryptedExtensions <= e.WriteCertificate &&
+		e.WriteCertificate <= e.WriteCertificateVerify &&
+		e.WriteCertificateVerify <= e.WriteServerFinished &&
+		e.WriteServerFinished <= e.ReadClientFinished
+}
+
+func (e EXP_EventTLS13ClientHandshakeTimingInfo) isMonotonic() bool {
+	return e.WriteClientHello <= e.ReadServerHello &&
+		e.ReadServerHello <= e.ReadEncryptedExtensions &&
+		e.ReadEncryptedExtensions <= e.ReadCertificate &&
+		e.ReadCertificate <= e.ReadCertificateVerify &&
+		e.ReadCertificateVerify <= e.ReadServerFinished &&
+		e.ReadServerFinished <= e.WriteClientFinished
+}
+
+type timingInfo struct {
+	serverTimingInfo EXP_EventTLS13ServerHandshakeTimingInfo
+	clientTimingInfo EXP_EventTLS13ClientHandshakeTimingInfo
+}
+
+func (r timingInfo) eventHandler(event EXP_Event) {
+	switch e := event.(type) {
+	case EXP_EventTLS13ServerHandshakeTimingInfo:
+		r.serverTimingInfo = e
+	case EXP_EventTLS13ClientHandshakeTimingInfo:
+		r.clientTimingInfo = e
+	}
+}
+
+func runHandshake(t *testing.T, clientConfig, serverConfig *Config) (timingState timingInfo, err error) {
+	const sentinel = "SENTINEL\n"
+	c, s := localPipe(t)
+	errChan := make(chan error)
+
+	clientConfig.EXP_EventHandler = timingState.eventHandler
+	serverConfig.EXP_EventHandler = timingState.eventHandler
+
+	go func() {
+		cli := Client(c, clientConfig)
+		err := cli.Handshake()
+		if err != nil {
+			errChan <- fmt.Errorf("client: %v", err)
+			c.Close()
+			return
+		}
+		defer cli.Close()
+		buf, err := ioutil.ReadAll(cli)
+		if err != nil {
+			t.Errorf("failed to call cli.Read: %v", err)
+		}
+		if got := string(buf); got != sentinel {
+			t.Errorf("read %q from TLS connection, but expected %q", got, sentinel)
+		}
+		errChan <- nil
+	}()
+
+	server := Server(s, serverConfig)
+	err = server.Handshake()
+	if err == nil {
+		if _, err := io.WriteString(server, sentinel); err != nil {
+			t.Errorf("failed to call server.Write: %v", err)
+		}
+		if err := server.Close(); err != nil {
+			t.Errorf("failed to call server.Close: %v", err)
+		}
+		err = <-errChan
+	} else {
+		s.Close()
+		<-errChan
+	}
+
+	return
+}
+
+func TestTLS13HandshakeTiming(t *testing.T) {
+	issuer, err := x509.ParseCertificate(testRSACertificateIssuer)
+	if err != nil {
+		panic(err)
+	}
+	rootCAs := x509.NewCertPool()
+	rootCAs.AddCert(issuer)
+
+	const serverName = "example.golang"
+
+	clientConfig := &Config{
```

Comment:
Since the `clientConfig` and `serverConfig` are nearly identical, you can start with a `baseConfig` then do `clientConfig := baseConfig.Clone()` and `severConfig := baseConfig.Clone()`. That'll save you some LOC.

Reply:
I didn't even know `Clone` was a thing. That works.